sandbox/gpio/common.go

```diff
@@ -0,0 +1,315 @@
+type ChardevChip struct {
+	Path     string
+	Name     string
+	Label    string
+	NumLines uint
+}
+
+func (c *ChardevChip) String() string {
+	return fmt.Sprintf("(name: %s, label: %s, lines: %d)", c.Name, c.Label, c.NumLines)
+}
+
+// This has to match the memory layout of `struct gpiochip_info` found
+// in /include/uapi/linux/gpio.h in the kernel.
+type kernelChipInfo struct {
+	name, label [32]byte
+	lines       uint32
+}
+
+const _GPIO_GET_CHIPINFO_IOCTL uintptr = 0x8044b401
+
+var unixSyscall = unix.Syscall
+
+var ioctlGetChipInfo = func(path string) (*kernelChipInfo, error) {
+	f, err := os.OpenFile(path, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	conn, err := f.SyscallConn()
+	if err != nil {
+		return nil, err
+	}
+
+	kci := new(kernelChipInfo)
+	var errno syscall.Errno
+	err = conn.Control(func(fd uintptr) {
+		_, _, errno = unixSyscall(unix.SYS_IOCTL, f.Fd(), _GPIO_GET_CHIPINFO_IOCTL, uintptr(unsafe.Pointer(kci)))
+	})
+	if errno != 0 {
+		return nil, errno
+	}
+	return kci, err
+}
+
+var chardevChipInfo = func(path string) (*ChardevChip, error) {
+	kci, err := ioctlGetChipInfo(path)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read gpio chip info from %q: %w", path, err)
+	}
+
+	chip := &ChardevChip{
+		Path:     path,
+		Name:     string(bytes.TrimRight(kci.name[:], "\x00")),
+		Label:    string(bytes.TrimRight(kci.label[:], "\x00")),
+		NumLines: uint(kci.lines),
+	}
+	return chip, nil
+}
+
+const (
+	aggregatorDriverDir        = "/sys/bus/platform/drivers/gpio-aggregator"
+	aggregatorNewDevicePath    = "/sys/bus/platform/drivers/gpio-aggregator/new_device"
+	aggregatorDeleteDevicePath = "/sys/bus/platform/drivers/gpio-aggregator/delete_device"
+	ephemeralUdevRulesDir      = "/run/udev/rules.d"
+)
+
+var lockAggregator = func() (unlocker func(), err error) {
+	flock, err := osutil.OpenExistingLockForReading(filepath.Join(dirs.GlobalRootDir, aggregatorDriverDir))
+	if err != nil {
+		return nil, err
+	}
+	if err := flock.Lock(); err != nil {
+		return nil, err
+	}
+	return func() {
+		// we don't care about the error, this is best-effort.
+		_ = flock.Close()
+	}, nil
+}
+
+var aggregatorCreationTimeout = 120 * time.Second
+
+func addAggregatedChip(ctx context.Context, sourceChip *ChardevChip, lines strutil.Range) (chip *ChardevChip, err error) {
+	// synchronize gpio helpers' access to the aggregator interface
+	unlocker, err := lockAggregator()
+	if err != nil {
+		return nil, err
+	}
+	defer unlocker()
+
+	watcher, err := inotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	defer watcher.Close()
+
+	err = watcher.AddWatch(dirs.DevDir, inotify.InCreate)
+	if err != nil {
+		return nil, err
+	}
+
+	// <label> <lines>
+	cmd := fmt.Sprintf("%s %s", sourceChip.Label, lines.String())
+	if err := os.WriteFile(filepath.Join(dirs.GlobalRootDir, aggregatorNewDevicePath), []byte(cmd), 0644); err != nil {
+		return nil, err
+	}
+
+	timeoutC := time.After(aggregatorCreationTimeout)
```

Comment:
Could you rewrite this with `time.NewTimer` to avoid the problem with how time.After behaves with regards to the garbage collector? A simple:
```go
timeoutTimer := time.NewTimer(aggregatorCreationTimeout)
defer timeoutTimer.Stop()
```

Then use `timeoutTimer.C` below.